### ISSUE TITLE
ibm_db attributes

### ISSUE THREAD

**Issue by lenhen** [ibmdb/python-ibmdb#345]
Hi,

I've installed python 3.6.5 on a RHEL server and then easy-installed ibm_db (2.0.9). After starting python I import ibm_db - which works fine - and when trying an ibm_db.connect I get the following error.

AttributeError: module 'ibm_db' has no attribute 'connect'

I then tried to list the attributes of ibm_db and it was a rather short list as compared with the documented attributes at https://github.com/ibmdb/python-ibmdb/tree/master/IBM_DB/ibm_db ...

It looked like this:

>>> import ibm_db
>>> dir(ibm_db)
['__doc__', '__loader__', '__name__', '__package__', '__path__', '__spec__']

What could have gone wrong here?
Thanks in advance, Len

**Comment by SabaKauser** [comment]
Hello,
ibm_db.connect() is an API.

**Comment by lenhen** [comment]
I see! So that's the reason it does not show up as a function :-)

However, the problem is that I cannot make ibm_db work as it is documented in examples.

I tried this:

conn_str='DATABASE=DB2X;HOSTNAME=db2x.fun.url.se;PORT=446;PROTOCOL=TCPIP;UID=NISSE;PWD=mypw;','',''
>>> ibm_db_conn = ibm_db.connect(conn_str)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'ibm_db' has no attribute 'connect'

As soon as I reference ibm_db.connect I get the above error message. So, I must do something completely wrong here. Why is python saying that the *attribute* connect does not exist? And how can I verify if it exist?
/Len

**Comment by SabaKauser** [comment]
Please follow documentation here -> https://github.com/ibmdb/python-ibmdb/wiki/APIs#ibm_dbconnect
Test samples can be found under : https://github.com/ibmdb/python-ibmdb/tree/master/IBM_DB/ibm_db/tests

The syntax of API is: ibm_db.connect(conn_str, '', '') for remote connection.
ibm_db.connect(database,userid, pwd) for local connection.

sanity test sample is provided in README.md here -> https://github.com/ibmdb/python-ibmdb/blob/master/IBM_DB/ibm_db/README.md

**Comment by kadler** [comment]
Do you have a Python file named ibm_db.py in your current directory? That's bitten me before when I create examples and accidentally name them the same as the module.

**Comment by lenhen** [comment]
Thanks for your involvement!

The only place where it is is in the following location.

$HOME/ibm_db/ibm_db-2.0.9/ibm_db.py

I've been trying again being more exact on the syntax, but whatever I do when calling the  ibm_db,connect API, I get the same message telling me that ibm_db has no attribute 'connect'

**Comment by kadler** [comment]
There is no "ibm_db.py" shipped by this project. There's an "ibm_db_dbi.py", but the ibm_db module is written in C. I think you have gotten an errant python file somehow.

**Comment by lenhen** [comment]
Interesting!
That is what was installed in my server when I invoked easy-install ibm_db and got the 2.0.9 version.
Here is what is found in the ibm_db-2.0.9 directory after the installation.
$ pwd
/usr/users/nisse/ibm_db/ibm_db-2.0.9
$ ls -l
total 1888
-rwxr-x---. 1 nisse shbgrp  12025 Aug  3 07:17 CHANGES
-rwxr-x---. 1 nisse shbgrp    547 Aug  6 17:20 config.py
-rwxr-x---. 1 nisse shbgrp  96256 Jul 30 10:34 ibm_db32_py27.pyd
-rwxr-x---. 1 nisse shbgrp  97792 Jul 27 11:33 ibm_db32_py33.pyd
-rwxr-x---. 1 nisse mygrp  96256 Jul 27 08:25 ibm_db32_py34.pyd
-rwxr-x---. 1 nisse mygrp 105984 Aug  3 05:31 ibm_db32_py35.pyd
-rwxr-x---. 1 nisse mygrp 105984 Aug  3 05:15 ibm_db32_py36.pyd
-rwxr-x---. 1 nisse mygrp 105984 Aug  3 05:48 ibm_db32_py37.pyd
-rwxr-x---. 1 nisse mygrp 125952 Jul 30 08:46 ibm_db64_py27.pyd
-rwxr-x---. 1 nisse mygrp 121856 Jul 27 11:05 ibm_db64_py33.pyd
-rwxr-x---. 1 nisse mygrp 120320 Jul 27 08:17 ibm_db64_py34.pyd
-rwxr-x---. 1 nisse mygrp 132096 Jul 31 07:20 ibm_db64_py35.pyd
-rwxr-x---. 1 nisse mygrp 132096 Jul 30 11:07 ibm_db64_py36.pyd
-rwxr-x---. 1 nisse mygrp 132096 Aug  3 07:00 ibm_db64_py37.pyd
-rwxr-x---. 1 nisse mygrp 360237 Aug  3 06:00 ibm_db.c
-rwxr-x---. 1 nisse mygrp  62448 Sep 14  2017 ibm_db_dbi.py
drwxr-x---. 2 nisse mygrp     90 Aug  3 07:55 ibm_db.egg-info
-rwxr-x---. 1 nisse mygrp  15686 Jul 23 05:52 ibm_db.h
-rwxr-x---. 1 nisse mygrp    478 Oct 26  2017 ibm_db.py
-rwxr-x---. 1 nisse mygrp   9164 Jul 23 05:52 LICENSE
-rwxr-x---. 1 nisse mygrp    180 Jul 23 05:52 MANIFEST.in
drwxr-x---. 2 nisse mygrp     35 Sep 26 17:28 __pycache__
-rwxr-x---. 1 nisse mygrp   9322 Aug  3 10:37 README.md
-rwxr-x---. 1 nisse mygrp  11565 Aug  3 08:28 setup.py
-rwxr-x---. 1 nisse mygrp   4220 Aug  3 05:35 testfunctions.py
drwxr-x---. 2 nisse mygrp   8192 Sep 24 16:24 tests
drwxr-x---. 2 nisse mygrp   8192 Aug  3 07:55 tests_1
-rwxr-x---. 1 nisse mygrp   1742 Jul 23 05:52 tests.py

That's what I got - and have not done any changes within this directory.

This is the content of the ibm_db.py file:
`import os
if 'clidriver' not in os.environ['PATH']:
    os.environ['PATH'] = os.environ['PATH'] + ";" + os.path.join(os.path.abspath(os.path.dirname(__file__)), 'clidriver', 'bin')
def __bootstrap__():
   global __bootstrap__, __loader__, __file__
   import sys, pkg_resources, imp
   __file__ = pkg_resources.resource_filename(__name__,'ibm_db_dlls\ibm_db.dll')
   __loader__ = None; del __bootstrap__, __loader__
   imp.load_dynamic(__name__,__file__)
__bootstrap__()
`

/Len

**Comment by lenhen** [comment]
If I mv the *ibm_db.py* file to *ibm_db.py.dontuse* and then try again after logging off and then on again. I still get the following message.

`AttributeError: module 'ibm_db' has no attribute 'connect'
`

/Len

**Comment by kadler** [comment]
The problem is that the ibm_db.py was installed in to site-packages instead of the ibm_db.so. It looks like the errant file is in the tar file on PyPI: https://files.pythonhosted.org/packages/3f/61/389c6decacfed61c2724005b8c2be61f2c699f731b53af13517e5a0e6e98/ibm_db-2.0.9.tar.gz

@SabaKauser looks like a new release needs to be uploaded to PyPI. I would suggest `2.0.9-1` as the version if you want to avoid all the `2.0.8a` issues.

**Comment by lenhen** [comment]
Thanks @kadler for your involvement in this issue and that you contribute to a solution!

/Len

**Comment by lenhen** [comment]
It's then obvious that *I've* missed something.

`Python 3.6.5 |Anaconda, Inc.| (default, Apr 29 2018, 16:14:56)
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ibm_db
>>> ibm_db.version
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'ibm_db' has no attribute 'version'
>>>
`
I will check and report back to you.

Thanks for your support!

/Len

**Comment by SabaKauser** [comment]
its ibm_db.__version__

**Comment by SabaKauser** [comment]
oops, sorry , git problem.
![image](https://user-images.githubusercontent.com/13583487/46192298-3031c680-c318-11e8-886c-cb93e01b84e2.png)
its version with two leading and trailing  underscores

**Comment by lenhen** [comment]
I checked my /usr/users/nisse/anaconda3/lib/python3.6/site-packages directory and I couldn't find the egg directory. I will look into this in more detail later today.

**Comment by kadler** [comment]
Huh, I didn't realize that when you install via `easy_install` or `setup.py install` it generates that python loader stub. I'm used to installing via pip, which just creates an ibm_db.so:

```
>>> import ibm_db
>>> ibm_db.__file__
'/home/kadler/venvs/db2-py2/lib/python2.7/site-packages/ibm_db.so'
```

```
>>> import ibm_db
>>> ibm_db.__file__
'/home/kadler/venvs/db2/lib/python3.6/site-packages/ibm_db-2.0.9-py3.6-linux-x86_64.egg/ibm_db.cpython-36m-x86_64-linux-gnu.so'
```

**Comment by lenhen** [comment]
And I get...

`>>> import ibm_db
>>> ibm_db.__file__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'ibm_db' has no attribute '__file__'`

...

**Comment by kadler** [comment]
@lenhen I'm not sure what that error is about. Were you ever able to get this working?

**Comment by lenhen** [comment]
Thanks for asking!

I just had to let it be while working on other stuff. But, I will have to 
restart the project soon. So will see if I can make it work then.

Thx, Len

Lennart Henäng   
_____________________________ 
Handelsbanken 
CDOC 
SE-115 82 STOCKHOLM 
Phone:  +46  8 701 8498
Mobile: +46 70 601 8498
e-mail: lehe08@handelsbanken.se



Från:   "Kevin Adler" <notifications@github.com>
Till:   "ibmdb/python-ibmdb" <python-ibmdb@noreply.github.com>
Kopia:  "lenhen" <lehe08@handelsbanken.se>, "Mention" 
<mention@noreply.github.com>
Datum:  2019-01-31 18:08
Ärende: Re: [ibmdb/python-ibmdb] ibm_db attributes (#345)



@lenhen I'm not sure what that error is about. Were you ever able to get 
this working?

**Comment by SabaKauser** [comment]
When you try next time, cd to your project directory. When you import ibm_db driver from where the ibm_db.py exist, python would import this module and it definitely does not have the python ibm_db APIs.

**Comment by timswiley** [comment]
I'm having this exact problem but i notice others too.   Running python 3.7 64bit and pip installed ibm_db successfully or so it says... However in visual studio code there are all kinds of complaints about it not recognizing many of the methods in the ibm_db import returning "has no *** member", however they do execute without raising any errors... i have code that calls ibm_db.connect and passes the is true test on being connected that i saw as a valid test for the connection.  I then prepare a simple sql statement "Select 1 as x From sysibm.sysdummy1" and perform exec_immediate.... NOTHING... it just leaves the program entirely... no error message or nothing it just exits the program after that line which is basically "ibm_db.exec_immediate(Con, 'Select 1 as x From sysibm.sysdummy1')  the user account has dba privileges (I'm the dba) so not permissions issue and i've tried several other sql statements to no avail... pyodbc works just fine 100%.  The reason i was trying the ibm_db class was pyodbc doesn't seem to like fetching BLOB/CLOB/DBCLOB data from db2 as it rejects that type as not support so i was hoping the ibm_db class would work... but i can't get past this very simple thing.   Side note... i've dba'd on Oracle/SQL Server/Informix/DB2/MySQL and programmed on all...  WTH does EVERYTHING seem to be harder or quirkier on DB2.... EVERYTHING....  (Sorry just venting)...

**Comment by SabaKauser** [comment]
@timswiley , As indicated earlier, the problem reported in this thread was with that of loading a python loader stub that is needed to load the ibm_db dll on windows and hence will report no attributes error.
I understand your feedback. 
I see  you have opened another issue and I will assist you there.

**Comment by srilamaiti** [comment]
I faced an interesting problem with ibm_db. The installed version in both prod and test environment is Python 3.6.8, installed thru Anaconda distribution. I have installed ibm-db, ibm-db-sa using pip in both environments. Prod ibm_db version is 3.0.0 and test version is 3.0.1. However, when I do the size comparison of ibm_db.cpython-36m-x86_64-linux-gnu.so between prod and test, i see it as 449K and 139K respectively. Any idea? Also, I have a custom library call in my code which invoking on it's own works fine in prod. Same thing fails in test environment with segmentation fault. If I precede the custom libarry call with import ibm_db, it goes thru. The libarry has ibm_db invocation as well. I am guessing it's something to do with ibm_db version and code. Any idea? Any pointer is much appreciated. Thanks in advance.
